### PR TITLE
added unix-specific API abstraction

### PR DESCRIPTION
**Pull request by hoopoepg**
as part of porting OFI into windows platform minor code refactoring
presented:
- removed all system-specific includes from public headers
- added unix specific sources (header & source), common for
  Linux/OSX/FreeBSD
- socket specific (and some other) calls are wrapped into
  abstraction layer & shifted into unix specific sources
- 'mix' macro is replaced by inline function (due to non-standard
  extention used in macro)

Signed-off-by: Oblomov, Sergey <sergey.oblomov@intel.com>

### CODE REVIEW

**Comment by lanl-ompi** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

File: include/rdma/fi_eq.h [hunk 1]
```diff
@@ -74,8 +72,8 @@ struct fid_wait {
 };
 
 struct fi_mutex_cond {
-	pthread_mutex_t		*mutex;
-	pthread_cond_t		*cond;
+	struct pthread_mutex_t	*mutex;
+	struct pthread_cond_t	*cond;
```
Comment (shefty) [inline include/rdma/fi_eq.h]: pthread_mutex_t and pthread_cond_t are data types, not structure names.  I admit I'm surprised this change didn't break the build.  But why was this made?

File: include/unix/osd.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+static inline int util_buf_memalign(void **memptr, size_t alignment, size_t size)
+{
+	return posix_memalign(memptr, alignment, size);
+}
+
+static inline void util_buf_freealign(void *memptr)
+{
+	free(memptr);
+}
```
Comment (shefty) [inline include/unix/osd.h]: Do we need to override this, versus just calling free?

File: include/unix/osd.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#ifndef _FI_UNIX_OSD_H_
+#define _FI_UNIX_OSD_H_
+
+#include <stdlib.h>
+#include <unistd.h>
+
+static inline int util_buf_memalign(void **memptr, size_t alignment, size_t size)
+{
+	return posix_memalign(memptr, alignment, size);
+}
```
Comment (shefty) [inline include/unix/osd.h]: please rename to ofi_memalign

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (hoopoepg) [inline include/rdma/fi_eq.h]: It is standard way to forward declaration of datatype. Actually usage of native OS specific type without wrapper is not best solution :)

File: include/unix/osd.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+static inline ssize_t fi_write_socket(int fd, const void *buf, size_t count)
+{
+	return write(fd, buf, count);
+}
+
+static inline int fi_close_socket(int socket)
+{
+	return close(socket);
+}
```
Comment (shefty) [inline include/unix/osd.h]: We have an effort to use the prefix 'ofi' for all internal calls/structures, and reserve the 'fi' prefix for external interfaces.  Please rename these using the ofi prefix.

File: include/unix/osd.h (hunk 2 above)
Comment (hoopoepg) [inline include/unix/osd.h]: On Windows platform there are pair functions to allocate and free aligned data

File: include/unix/osd.h (hunk 4 above)
Comment (hoopoepg) [inline include/unix/osd.h]: Sure, will do tomorrow
Thank you

**Comment by shefty** [comment]
@j-xiong @hppritcha @jithinjosepkl @goodell @bturrubiates - please review

These changes look reasonable to me.  I'm not thrilled with the _socket suffix on the function names, but I don't have a better idea, other than using send/recv instead of read/write.

File: include/unix/osd.h (hunk 4 above)
Comment (shefty) [inline include/unix/osd.h]: The rename can even come as a separate patch that follows these if it makes it easier.

Comment (hoopoepg) [inline include/unix/osd.h]: No problemo :)
Vim make it easy

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (bturrubiates) [inline include/rdma/fi_eq.h]: > pthread_mutex_t and pthread_cond_t are data types, not structure names. I admit I'm surprised this change didn't break the build. But why was this made?

It doesn't break the build because of the casts in `sock_wait.c`. 

> It is standard way to forward declaration of datatype. Actually usage of native OS specific type without wrapper is not best solution :)

I don't see why this is necessary, `pthread_mutex_t` are typedefs that work across all platforms.

Comment (hoopoepg) [inline include/rdma/fi_eq.h]: what about windows?

Comment (j-xiong) [inline include/rdma/fi_eq.h]: There are ways to use pthread on windows and that's probably the easiest way to get current code working on windows. However, if Window's native threading API is to be used (and wrapped in pthread API),  it would be less confusing with defining these types directly(instead of "struct") in the os dependent headers.

Comment (hoopoepg) [inline include/rdma/fi_eq.h]: right, but in this case we have to add construct like 
```
#if Linux
use-pthread-types
#else
use-windows-threading
#endif
```
and same for includes of interface headers (pthread or windows.h). previous PR was declined due to too many if-defs :)

File: include/rdma/fi_log.h [hunk 5]
```diff
@@ -35,8 +35,6 @@
 #ifndef FI_LOG_H
 #define FI_LOG_H
 
-#include "config.h"
```
Comment (bturrubiates) [inline include/rdma/fi_log.h]: This needs to be included here, `ENABLE_DEBUG` is used lower in the file. This will break logging.

Comment (hoopoepg) [inline include/rdma/fi_log.h]: really bad idea: config.h is widely used header which is automatically generated by some autotools. as i can see config.h is not exported on 'make install', and it may conflict with user's configuration

or i missed something?

**Comment by j-xiong** [comment]
Possible alternative to the _socket suffix is _fd, so we have "ofi_open_fd", etc, or even change it to "ofi_fd_open" to be consistent with "ofi_fd_nonblock". Non necessarily better, just some ideas to consider.

File: include/rdma/fi_log.h (hunk 5 above)
Comment (bturrubiates) [inline include/rdma/fi_log.h]: > really bad idea: config.h is widely used header which is automatically generated by some autotools. as i can see config.h is not exported on 'make install', and it may conflict with user's configuration
> 
> or i missed something?

It doesn't need to be exported on `make install`, this `config.h` is created by autotools and one of the things it defines is whether it is a debug build or production build. The file is included at compile time. 

How would it conflict with a users configuration?

Comment (hoopoepg) [inline include/rdma/fi_log.h]: user's configuration may be generated not by autotools, and ENABLE_DEBUG may be omit there (and config.h as well), in this case compiltion will be aborted on user's side

Comment (hoopoepg) [inline include/rdma/fi_log.h]: BTW, to workaround warning in fi_log construct may be used:
`#if defined ENABLE_DEBUG && ENABLE_DEBUG > 0`

Comment (bturrubiates) [inline include/rdma/fi_log.h]: > user's configuration may be generated not by autotools, and ENABLE_DEBUG may be omit there (and config.h as well), in this case compiltion will be aborted on user's side

What do you mean by "user's configuration"? This config.h file is always generated by autotools. If this file is there, then it was generated by auto tools and `ENABLE_DEBUG` is defined one way or the other. The `config.h` file is not a general configuration file that a user can define.

Comment (hoopoepg) [inline include/rdma/fi_log.h]: cmake/qmake/no autotools on user's side, scons/what else? :)

on Linux there are tons of build systems :)

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (j-xiong) [inline include/rdma/fi_eq.h]: I don't understand here. If you define pthread wrappers and pthread_xxxx_t types in "windows/osd.h", no changes are needed outside the "windows" directory.

In addition, how changing the parameter types in the prototype to "struct pthread_xxx_t" makes life any easier, not considering the confusion it may bring?

File: include/rdma/fi_log.h (hunk 5 above)
Comment (bturrubiates) [inline include/rdma/fi_log.h]: > BTW, to workaround warning in fi_log construct may be used:
> #if defined ENABLE_DEBUG && ENABLE_DEBUG > 0

In this situation where we don't have this file, is it because of using a non-autotools based build system (perhaps because Windows)?

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (hoopoepg) [inline include/rdma/fi_eq.h]: actually it allows compilation of user's code without including OS specific runtime (it may be compiled without pthread headers on Linux). osd.h headers are not part of external headers and can't be used in rdma/* files

File: include/rdma/fi_log.h (hunk 5 above)
Comment (hoopoepg) [inline include/rdma/fi_log.h]: why not? in case if user includes his config.h (or any other configuration header) prior to fi_log - then user will have standard way to enable debugging. alternative way is control ENABLE_DEBUG by command line (or other configuration options) to enable debuggins. in any case user is free to enable such option as he want

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (j-xiong) [inline include/rdma/fi_eq.h]: If the reason is osd.h not being part of external headers, we may need to consider either move them to the external headers or remove os dependent references from rdma/*.

Comment (hoopoepg) [inline include/rdma/fi_eq.h]: totally agree, but changing of this interface may break backward compatibility

File: include/unix/osd.h (hunk 4 above)
Comment (hoopoepg) [inline include/unix/osd.h]: done, renamed to ofi_*****

File: src/fasthash.c [hunk 6]
```diff
@@ -27,10 +27,13 @@
 
 // Compression function for Merkle-Damgard construction.
 // This function is generated using the framework provided.
-#define mix(h) ({					\
-			(h) ^= (h) >> 23;		\
-			(h) *= 0x2127599bf4325c37ULL;	\
-			(h) ^= (h) >> 47; })
+static inline uint64_t mix(uint64_t h)
+{ /* use inline function instead of non-standard macro */
```
Comment (goodell) [inline src/fasthash.c]: Let's kill this comment.  After the change has been made, it seems unlikely that anyone will go out of their way to put the GCC extension back in place even if the comment is not present.

File: src/unix/osd.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2015 Los Alamos Nat. Security, LLC. All rights reserved.
```
Comment (goodell) [inline src/unix/osd.c]: Missing Intel copyright addition here? (and possibly on some of the other files you modified)

File: include/rdma/fi_log.h (hunk 5 above)
Comment (jsquyres) [inline include/rdma/fi_log.h]: Note that `fi_log.h` is not a user-visible file -- it's for internal use inside libfabric only (perhaps it should not be installed by `make install`...?).

Hence, it is totally fine to it to use the private `config.h` that is generated by the GNU Autotools.  There will never be the possibility of an application *using* libfabric conflicting with our `config.h` because it's private and not visible to libfabric-using applications.

We do not support any other build system on Linux/OS X -- any road that tries to support multiple systems simultaneously (cmake, autotools, scons, ...etc.) when you don't have to leads to madness.  Supporting Visual Studio in addition to GNU Autotools is going to be bad enough (personally, I'm of the opinion that the GNU Autotools build system is the "master" build system, and the Visual Studio support will need to keep up with whatever changes are made in the GNU Autotools build system -- such as providing a dummy `config.h` file), but there's currently not much of an option if native Windows support is desired.

Adding support for a 3rd build system would be... undesirable.

**Comment by jsquyres** [comment]
Please also see https://github.com/ofiwg/libfabric/issues/2044.

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (shefty) [inline include/rdma/fi_eq.h]: Since these are user facing header files, we may need to treat those changes as a special case.  I would rather not include osd.h from a user header file if we can help it.  Maybe this section of code is ifdef out for windows.  It can then be defined in the windows osd.h file, so that all internal code that relies of the definition will work

File: include/rdma/fi_log.h (hunk 5 above)
Comment (shefty) [inline include/rdma/fi_log.h]: fi_log.h is user-visible.  It's just in this case that the user is an external provider.  The config.h file really shouldn't be included from fi_log.h.

**Comment by shefty** [comment]
@j-xiong The _fd suffix doesn't quite work.  On windows the calls only work with sockets, which are not fd's.  So the calls need to indicate that only socket fd's should be passed to the call.

File: include/rdma/fi_log.h (hunk 5 above)
Comment (jsquyres) [inline include/rdma/fi_log.h]: `fi_log.h` is not documented as being user-visible; that's why I concluded that it is not user-visible.  If it *is* supposed to be user-visible, then I agree: it should not include `config.h` -- there are standard workarounds for getting configury output in user-visible header files (e.g., OMPI does this kind of thing).

Is it an explicit goal to be able to build providers out-of-tree?  If so, is it known if that actually works?

Comment (bturrubiates) [inline include/rdma/fi_log.h]: > fi_log.h is user-visible. It's just in this case that the user is an external provider. The config.h file really shouldn't be included from fi_log.h.

I see the confusion here now, Jeff and I both thought that `fi_log.h` was internal only.

Comment (shefty) [inline include/rdma/fi_log.h]: All files in include/rdma are installed and visible.  fi_log and fi_prov are usable by external providers.  It is a goal to support out of tree providers, so that a single provider can be updated without re-compiling the entire library and replacing the build-in providers.  I'm guessing that we can support that by building a dynamic provider in-tree and only installing the provider library, if we didn't want to allow building the provider out of tree.

I'm not aware of any providers being built out of tree.

File: prov/gni/src/gnix_wait.c [hunk 8]
```diff
@@ -277,8 +274,8 @@ int gnix_wait_close(struct fid *wait)
 	}
 
 	if (wait_priv->type == FI_WAIT_FD) {
-		close(wait_priv->fd[WAIT_READ]);
-		close(wait_priv->fd[WAIT_WRITE]);
+		ofi_close_socket(wait_priv->fd[WAIT_READ]);
+		ofi_close_socket(wait_priv->fd[WAIT_WRITE]);
 	}
 
 	_gnix_ref_put(wait_priv->fabric);
```
Comment (shefty) [inline prov/gni/src/gnix_wait.c]: please revert changes to this file.  The gni provider won't support windows anyway.

File: prov/usnic/src/usdf_pep.c [hunk 9]
```diff
@@ -581,16 +582,11 @@ usdf_pep_open(struct fid_fabric *fabric, struct fi_info *info,
...
-                goto fail;
-        }
+	ret = fi_fd_nonblock(pep->pep_sock);
+	if (ret) {
+		ret = -errno;
+		goto fail;
+	}
 
 	/* set SO_REUSEADDR to prevent annoying "Address already in use" errors
 	 * on successive runs of programs listening on a well known port */
```
Comment (shefty) [inline prov/usnic/src/usdf_pep.c]: Even though usnic won't support windows, please keep these changes, as they're good anyway.  Pending approval from Cisco - @goodell

File: include/rdma/fi_log.h (hunk 5 above)
Comment (bturrubiates) [inline include/rdma/fi_log.h]: > All files in include/rdma are installed and visible. fi_log and fi_prov are usable by external providers. It is a goal to support out of tree providers, so that a single provider can be updated without re-compiling the entire library and replacing the build-in providers. I'm guessing that we can support that by building a dynamic provider in-tree and only installing the provider library, if we didn't want to allow building the provider out of tree.
> 
> I'm not aware of any providers being built out of tree.

Isn't the A3CUBE provider out of tree?

File: prov/usnic/src/usdf_pep.c (hunk 9 above)
Comment (goodell) [inline prov/usnic/src/usdf_pep.c]: Yes, in this case, please keep these changes to the `usnic` provider.

File: include/rdma/fi_log.h (hunk 5 above)
Comment (jsquyres) [inline include/rdma/fi_log.h]: If there's no desire to build an out-of-tree provider, then there's no need to install `fi_log.h` (during `make install`).

File: src/unix/osd.c (hunk 7 above)
Comment (hoopoepg) [inline src/unix/osd.c]: hmmm... i just copied file from linux directory & added fi_nonblock function... i will update copyright notice

File: include/fi_rbuf.h [hunk 10]
```diff
@@ -226,16 +227,15 @@ static inline int rbfdinit(struct ringbuffd *rbfd, size_t size)
 	if (ret < 0)
 		goto err1;
 
-	flags = fcntl(rbfd->fd[RB_READ_FD], F_GETFL, 0);
-	ret = fcntl(rbfd->fd[RB_READ_FD], F_SETFL, flags | O_NONBLOCK);
-	if (ret < 0)
+	ret = fi_fd_nonblock(rbfd->fd[RB_READ_FD]);
+	if (ret)
```
Comment (jithinjosepkl) [inline include/fi_rbuf.h]: Shouldnt this be ofi_ too?

Comment (hoopoepg) [inline include/fi_rbuf.h]: no ideas. i just replaced 2*fcntl by call to existing function :)
but you are right, fi_fd_... is out of naming convention

File: prov/sockets/src/sock_conn.c [hunk 11]
```diff
@@ -467,8 +466,10 @@ struct sock_conn *sock_ep_connect(struct sock_ep_attr *ep_attr, fi_addr_t index)
 	if (!do_retry)
 		goto err;
 
-	close(conn_fd);
-	conn_fd = -1;
+	if (conn_fd != -1) {
+		ofi_close_socket(conn_fd);
+		conn_fd = -1;
+	}
```
Comment (jithinjosepkl) [inline prov/sockets/src/sock_conn.c]: Why do we need the check for -1 here?

File: prov/sockets/src/sock_progress.c [hunk 12]
```diff
@@ -2466,6 +2466,8 @@ static int sock_pe_progress_rx_ep(struct sock_pe *pe, struct sock_ep_attr *ep_at
 	fastlock_acquire(&map->lock);
 	for (i = 0; i < num_fds; i++) {
 		fd = sock_epoll_get_fd_at_index(&map->epoll_set, i);
+		if(fd == -1) /* failed to lookup fd due to connection failures */
+			continue;
```
Comment (jithinjosepkl) [inline prov/sockets/src/sock_progress.c]: Do we need the check here? Shouldn't this be handled by the idm_lookup?

Comment (hoopoepg) [inline prov/sockets/src/sock_progress.c]: as i can see -1 is returned on windows platform only (there is bit different socket state handling)

File: prov/sockets/src/sock_conn.c (hunk 11 above)
Comment (hoopoepg) [inline prov/sockets/src/sock_conn.c]: during porting to windows i catched situation when conn_fd was not properly processed & has value -1

File: include/fi_rbuf.h (hunk 10 above)
Comment (shefty) [inline include/fi_rbuf.h]: We are in the process of using the ofi prefix for internal calls, but most still use fi.  I think the change here is fine, since it's just using an existing call.  I only asked that new calls use the newer prefix.  Updating the existing calls is really a separate issue.

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (shefty) [inline include/rdma/fi_eq.h]: To move forward on this patch, please revert this change and instead use an ifdef to remove it from the windows build.

File: include/rdma/fi_log.h (hunk 5 above)
Comment (shefty) [inline include/rdma/fi_log.h]: To move forward with this patch, please revert this change until we figure out what to do with this file.

File: include/rdma/fi_log.h [hunk 13]
```diff
@@ -87,7 +85,7 @@ void fi_log(const struct fi_provider *prov, enum fi_log_level level,
 #define FI_INFO(prov, subsystem, ...)					\
 	FI_LOG(prov, FI_LOG_INFO, subsystem, __VA_ARGS__)
 
-#if ENABLE_DEBUG
+#if defined ENABLE_DEBUG && ENABLE_DEBUG > 0
```
Comment (shefty) [inline include/rdma/fi_log.h]: To move forward with this patch, please revert this change.  If you want, please submit the changes to fi_log.h as a separate commit

File: prov/sockets/src/sock_wait.c [hunk 14]
```diff
@@ -62,8 +62,8 @@ int sock_wait_get_obj(struct fid_wait *fid, void *arg)
 		break;
 
 	case FI_WAIT_MUTEX_COND:
-		mut_cond.mutex = &wait->wobj.mutex_cond.mutex;
-		mut_cond.cond = &wait->wobj.mutex_cond.cond;
+		mut_cond.mutex = (struct pthread_mutex_t*)&wait->wobj.mutex_cond.mutex;
+		mut_cond.cond = (struct pthread_cond_t*)&wait->wobj.mutex_cond.cond;
```
Comment (shefty) [inline prov/sockets/src/sock_wait.c]: please revert this change

**Comment by shefty** [comment]
There's been a lot of discussion on this patch.  I marked a few changes to revert, plus one change request.  If you can make those updates, I think we can move forward on merging the other changes.  Then we can start on the next round of updates.

File: prov/sockets/src/sock_wait.c (hunk 14 above)
Comment (hoopoepg) [inline prov/sockets/src/sock_wait.c]: in this case there is compilation warning, because ptherad types were declared 'forwarded'

File: include/rdma/fi_eq.h (hunk 1 above)
Comment (hoopoepg) [inline include/rdma/fi_eq.h]: i applied another type of forward declaration:
```
struct pthread_mutex_t;
struct pthread_cond_t;

struct fi_mutex_cond {
       pthread_mutex_t>*mutex;
       pthread_cond_t> *cond;
};
```
which better fit to C standard. patch to sock_wait.c is reverted

File: prov/sockets/src/sock_wait.c (hunk 14 above)
Comment (hoopoepg) [inline prov/sockets/src/sock_wait.c]: done.

**Comment by hoopoepg** [comment]
done, 2 changes are reverted

File: include/rdma/fi_eq.h [hunk 15]
```diff
@@ -73,9 +71,12 @@ struct fid_wait {
 	struct fi_ops_wait	*ops;
 };
 
+struct pthread_mutex_t;
+struct pthread_cond_t;
```
Comment (goodell) [inline include/rdma/fi_eq.h]: Don't do this.  The `struct pthread_XXX_t` form is **nonstandard** and we should not be messing with their namespace.  Leave the `#include <pthread.h>` and guard it with an `#if !WINDOWS`, [per Sean's comment earlier](https://github.com/ofiwg/libfabric/pull/2041#discussion_r63965793).

Similarly, put an `#if !WINDOWS` around the members of the `struct fi_mutex_cond` below.  When it comes time to add Windows support properly, we can look at whether to put Windows style mutexes here with an #ifdef or defining an additional comparable type for the Windows platform (`struct fi_win_mutex_cond` or whatever).

File: include/rdma/fi_eq.h [hunk 16]
```diff
@@ -73,9 +71,12 @@ struct fid_wait {
...
+
 struct fi_mutex_cond {
-	pthread_mutex_t		*mutex;
-	pthread_cond_t		*cond;
+	pthread_mutex_t	*mutex;
+	pthread_cond_t	*cond;
```
Comment (goodell) [inline include/rdma/fi_eq.h]: And revert this whitespace change, it was done needlessly.

File: include/rdma/fi_eq.h (hunk 15 above)
Comment (hoopoepg) [inline include/rdma/fi_eq.h]: non-standard? sure? i tested such construct on all available compilers (icc/gcc/MS) - it works everywhere...
i will revert these changes

Comment (goodell) [inline include/rdma/fi_eq.h]: My concern has never been about whether this change _worked_, but whether it is understandable and _appropriate_, especially in libfabric's public headers.  Thanks for reverting it.

**Comment by shefty** [comment]
The only remaining objection to this patch is the removal of config.h from fi_log.h.  Please revert that.  Then I think we can merge what's left and continue the discussion around how to handle fi_log.h and fi_prov.h.

**Comment by hoopoepg** [comment]
fi_log.h reverted

**Comment by shefty** [comment]
thanks

**Comment by hoopoepg** [comment]
Great!!!
Thank you

**Comment by hppritcha** [comment]
@hoopoepg please note you must have done a merge rather than a rebase at some point.  Hence my 
commit bf31383 mysteriously appearing twice.

You might want to review the difference between rebasing your topic branch on top of master as opposed to merging.  This time it was okay - I think - but this could potentially cause problems in with future PRs.
